### ISSUE TITLE
User - Account - Login via Facebook

### ISSUE THREAD

**Issue by Mcscott142** [revelrylabs/bn-mobile-react#150]
requires https://github.com/big-neon/bn-api-node/issues/322

## Background
We'll want to make it easy for users to access/login to the application. One way to do this is allowing them to login via Facebook. 

## Scenario:  A user opens the Big Neon App
- [ ] When  I select login with Facebook
- [ ] Then I am signed into the application using my Facebook login

**Comment by Mcscott142** [comment]
## Background
We'll want to make it easy for users to access/login to the application. One way to do this is allowing them to login via Facebook. 

## Scenario:  A user opens the Big Neon App
- [ ] When  I select login with Facebook
- [ ] Then I am signed into the application using my Facebook login

**Comment by jwietelmann** [comment]
https://docs.expo.io/versions/latest/sdk/facebook

**Comment by jwietelmann** [comment]
@Krakaw or @zmelamed Can you please check on whatever FB app stuff you already have set up, make sure it has everything from [this doc](https://docs.expo.io/versions/latest/sdk/facebook/), and then send me the bits I need to be able to wire it up?

**Comment by Mcscott142** [comment]
@zmelamed Looks like what we need for iOS is facebookScheme, Facebook App Id, and Display Name:

>Add your app's Bundle ID as a Bundle ID in app's settings page pictured above.

>In your app.json, add a field facebookScheme with your Facebook login redirect URL scheme found here under 4. Configure Your info.plist. It should look like "fb123456". - https://developers.facebook.com/docs/facebook-login/ios

>Also in your app.json, add your Facebook App ID and Facebook Display Name under the facebookAppId and facebookDisplayName keys.

and `Facebook Key Hash` for Android: 

> Copy Facebook Key Hash and paste it as an additional key hash in your Facebook developer page pictured above.

We can implement once we've got the right info, you can add @daybreaker and @jwietelmann as Developers to the Big Neon Facebook app also, and we could probably get the info ourselves.

**Comment by jwietelmann** [comment]
@Mcscott142 it is important that whoever does this actually read that doc page I linked. We don't just need info out of it. We need some things to be done in the FB console, per that guide.

**Comment by zmelamed** [comment]
ok I think it should all be configured properly:
`facebookScheme `:`fb1737006886375028`
`facebookAppId`:`1737006886375028`
`facebookDisplayName`: `Big Neon`

If you guys give me your facebook IDs I'll get you added as devs as well

**Comment by jwietelmann** [comment]
@zmelamed mine is `joel.wietelmann` (couldn't find the numeric ID because they always change stuff around, but that string should work IIRC)

**Comment by Mcscott142** [comment]
@daybreaker What's your FB ID? @zmelamed mine is 27433687

**Comment by jwietelmann** [comment]
Hypothetically we could begin work on this now before we're added as FB app devs, at least until we find or don't find that everything is set up to work properly.

**Comment by brittanygay** [comment]
@daybreaker let me know if this will need any UX tweaking, I remember having it set up but we commented it out. If any design broke or looks off just let me know.

**Comment by zmelamed** [comment]
I want to make sure we know not to include this in a deploy until we're also ready to deploy it on the web, so it will require coordination. We can't have people creating accounts with facebook unless they can use that auth method in both places.

**Comment by daybreaker** [comment]
Had a lot of issues generating an Android FB key. Somehow the android keystore got corrupted with the wrong password. I was able to clear it and rebuild it though.

**Comment by daybreaker** [comment]
On hold until the web interface is done.

**Comment by Mcscott142** [comment]
@Krakaw I guess we need to sync up on rolling this out. It is ready for Web now?

**Comment by jwietelmann** [comment]
I'm going to leave this blocked and up next, but let it be known that I'm going to try and see how much working code I can get in there for when it's actually unblocked.

**Comment by jwietelmann** [comment]
I've done some of the setup work, but we are still blocked from progress by the following problems:

* Lack of Big Neon API support (https://github.com/big-neon/bn-api-node/issues/322)
* We're not yet registered test users in Facebook
![Image from iOS (1).png](https://waffleio-direct-uploads-production.s3.amazonaws.com/uploads/53b5aa8a7953d2fd0f72338d/125516c66e82c728ace21e0d46c1999b2487c3af9de6b203ad72c3d461176e316546b3397f8fe62bab0021435a0d48a8470e5840a8a88d2cafe33c708b0e03edd16f5874d42103eee9dc4be05ce36ade0d97bcae8b4bcae93820e58e86463a16011c80f84f78.png)

**Comment by Krakaw** [comment]
@jwietelmann, @daybreaker and @Mcscott142 have been added, please send me your details so I can add you too.

**Comment by jwietelmann** [comment]
@Krakaw My unique string is `joel.wietelmann` and my numeric is `851505030363`

**Comment by Krakaw** [comment]
Thanks, I've added you. 
Is this the endpoint you are requiring? https://big-neon.github.io/bn-api-node/classes/_resources_external_.externalresource.html#facebooklogin

**Comment by jwietelmann** [comment]
Seems like it. I’ll give it a go and let you know if I have any problems.

**Comment by jwietelmann** [comment]
@Krakaw This is what I get from the native Facebook integration:

```
[10:32:02] Object {
[10:32:02]   "declinedPermissions": Array [],
[10:32:02]   "expires": 1557844201,
[10:32:02]   "permissions": Array [
[10:32:02]     "email",
[10:32:02]     "public_profile",
[10:32:02]   ],
[10:32:02]   "token":  [REDACTED],
[10:32:02] }
```

This is what the endpoint requires:

```
Error: /external/facebook/web_login is expecting these fields: accessToken, userID, expiresIn, signedRequest
```

I'm not sure what `signedRequest` is supposed to be, and also it probably makes more sense to have the BN API look up the user ID with this: https://graph.facebook.com/me?fields=id&access_token=MY_ACCESS_TOKEN

(Also, I'm not 100% sure what will happen if I deny BN access to my email address, but I suspect it will be a problem.)

**Comment by jwietelmann** [comment]
Putting this back down because I'm hard-blocked on the `signedRequest` thing.

**Comment by Krakaw** [comment]
@jwietelmann we are using https://connect.facebook.net/en_US/sdk.js and have implemented it here: https://github.com/big-neon/bn-web/blob/65958e6f7af057b0212409a97c5a91336edafc21/src/components/pages/authentication/social/FacebookButton.js#L87

**Comment by Mcscott142** [comment]
This one should be ready to work.

**Comment by jwietelmann** [comment]
Our implementation uses the _native integration with the Facebook mobile app_, which is much better for the user. I don't think that trying to switch to the FB JS SDK auth response is a suitable alternative.

**Comment by Krakaw** [comment]
I've removed the required fields that were blocking you, it only requires the accessToken field as of version 1.0.12 https://big-neon.github.io/bn-api-node/classes/_resources_external_.externalresource.html#facebooklogin

**Comment by Mcscott142** [comment]
Unblocking cc @bryanjos @daybreaker

**Comment by bryanjos** [comment]
Still blocked for me because I don't have access. 

My facebook ID is `bryanjos`
numberic ID `45700686`

**Comment by bryanjos** [comment]
cc @Krakaw or @zmelamed ^

**Comment by daybreaker** [comment]
@Krakaw is the field still required in the main API?

still getting `{"error":"missing field `signedRequest` at line 1 column 293"}` as the response from bn-node-api

**Comment by Krakaw** [comment]
Let me double check, to bypass that just include a blank string for now but I'll confirm and fix it.

**Comment by daybreaker** [comment]
Thanks! Returning an access & refresh token now!

**Comment by Mcscott142** [comment]
I'm not able to login via iOS Facebook App when testing via Expo. If using the safari browser then I'm able to. I have not tested on Android yet

Repro Steps:
1) Select Facebook Connect
2) Select Continue on pop out when asked to use facebook to sign in 
2) Select continue with App via Safari page
3) Select "Open Facebook"
4) Select Continue from Facebook app when asked if I'd like to continue
5) Screen stays on Facebook App and the screen is blank

<details>
<summary>iOS Screenshot</summary>

![Image from iOS (7)](https://user-images.githubusercontent.com/6877299/54690208-e370c080-4aee-11e9-8f58-dfbe1e78ee17.png)

</details>

**Comment by Mcscott142** [comment]
@daybreaker see above comment https://github.com/revelrylabs/bn-mobile-react/issues/150#issuecomment-474842749

**Comment by Mcscott142** [comment]
Working for me on latest stage build (.40) Moving to UAT
